{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TARA oceans MAGs\n",
    "\n",
    "1. https://www.genoscope.cns.fr/tara/\n",
    "\n",
    "Original study: https://www.sciencedirect.com/science/article/pii/S2666979X22000477"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MedDB database of marine eukaryotic genomes\n",
    "\n",
    "https://metdb.sb-roscoff.fr/metdb/"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Experiments on phycospheres\n",
    "\n",
    "1. https://www.nature.com/articles/nature14488\n",
    "2. https://www.nature.com/articles/ismej2017117\n",
    "\n",
    "3. Taxonomy of HAB: https://journals.asm.org/doi/10.1128/AEM.00349-19?url_ver=Z39.88-2003&rfr_id=ori:rid:crossref.org&rfr_dat=cr_pub%20%200pubmed"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Modeling the phycosphere of a HAM\n",
    "\n",
    "\n",
    "Refs:\n",
    "\n",
    "1. https://www.nature.com/articles/s41598-021-81082-1\n",
    "\n",
    "\n",
    "2. Bacterial community dynamics influeced by type of secreted substrate during microalgal life cycle: https://www.sciencedirect.com/science/article/pii/S0043135420300907"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fake phycosphere MAGs\n",
    "\n",
    "1. https://journals.asm.org/doi/10.1128/MRA.00724-20"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## List of marine harmful algal bloom species\n",
    "\n",
    "1. https://www.marinespecies.org/hab/\n",
    "\n",
    "2. https://www.sciencedirect.com/science/article/pii/S1568988320301220  (Mediterranean Sea)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MAGs of Atlantic phycospheres\n",
    "\n",
    "1. https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-022-01254-7"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Studies on HAB events\n",
    "\n",
    "1. https://www.sciencedirect.com/science/article/pii/S1568988320301220  (Mediterranean Sea)\n",
    "2. https://www.nature.com/articles/ismej2013138 (Pseudo-nitzschia)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Previous modelling of HAB events\n",
    "\n",
    "1. https://www.pnas.org/doi/10.1073/pnas.1619575114"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Community modeling\n",
    "\n",
    "1. Review: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8202304/\n",
    "2. MetaT-guided genome-scale modeling of communities: https://www.sciencedirect.com/science/article/pii/S2667237522002880?via%3Dihub"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HAB phycosphere modeling workplan\n",
    "\n",
    "1. Select candidate HAB species from literature\n",
    "2. Search MAGs for these species in TARA, ideally from the same metagenome.\n",
    "3. Build GEMs \n",
    "4. Build community models\n",
    "\n",
    "\n",
    "### Candidate HAB species\n",
    "\n",
    "__Dinoflagellates__:\n",
    "\n",
    "1. Gymnodinium catenatum: https://www.sciencedirect.com/science/article/pii/S0944501318306189 (samples in China)\n",
    "   Bacterial community 4 most abundant orders: Gammaproteobacteria, Alphaproteobacteria, Synechococcophycideae, Flavobacteriia\n",
    "   Bacterial community 4 most abundant classes: Oceanospirillales, Rhodobacterales, Synechoococcales, Flavobacteriales\n",
    "\n",
    "\n",
    "__Diatoms__:\n",
    "\n",
    "Pseudo-nitzschia causing: domoic acid which causes Amnesic Shellfish Poisoning (ASP) also called Domoic Acid Poisoning (DAP). [REF](https://isdr.org/diatom-of-the-month-august-2019-diatom-blooms-and-their-effects-on-human-health/)\n",
    "\n",
    "\n",
    "\"DA, which poses a threat to ocean and human health, also brings monetary losses from temporary closure of shellfish aquaculture, seafood and tourism industries worldwide (Hoagland et al., 2002; Trainer et al., 2007). Human intoxication is referred to as amnesic shellfish poisoning; clinical signs in humans include gastrointestinal distress, confusion, disorientation, seizures, permanent short-term memory loss and in severe cases, death (Perl et al., 1990).\"\n",
    "\n",
    "\"leading to hypotheses about the role of DA in modulating the associated heterotrophic bacterial assemblage (Bates et al., 1995; Guannel et al., 2011; Trainer et al., 2012).\"\n",
    "\n",
    "\"In general, we found that foreign microbiota enhances production of DA of the toxigenic Pseudo-nitzschia. P. australis DA production range from 38.2–1398.5 femtogram per cell (or 44.7–823.7 μg l−1) depending on the type of microbiota it harbors\"\n",
    "\n",
    "\"P. australis grown with the Gamma-proteobacteria, PP–Marinobacter and PP–Alteromonas showed significantly higher SPG than axenic cultures but not with PP–Pseudoalteromonas\"\n",
    "\n",
    "\"PP–Polaribacter also significantly increased the SPG of P. australis \"\n",
    "\n",
    "\"P. australis grown with the Gamma-proteobacteria, PP–Marinobacter and PP–Alteromonas showed significantly higher SPG than axenic cultures but not with PP–Pseudoalteromonas \"\n",
    "\n",
    "[REF](https://www.nature.com/articles/ismej2013138)\n",
    "\n",
    "\n",
    "DA: Domoic Acid\n",
    "SPG: Specific Algal Growth Rate\n",
    "\n",
    "2. Pseudo-nitzschia: https://www.nature.com/articles/ismej2013138 (pseudo-nitzschia microbiome)\n",
    "\n",
    "These species increased DA content in P. australis:\n",
    " PP–Alteromonas\n",
    " PP–Glaciecola\n",
    " PP–Polaribacter \n",
    "\n",
    "These species decreased DA content in P. australis:\n",
    "PP-Roseobacter\n",
    "\n",
    "\n",
    "### Increased Pseudo-Nitzschia HABs in the Alskan arctic due to ocean warming\n",
    "\n",
    "https://tos.org/oceanography/article/harmful-algal-blooms-in-the-alaskan-arctic-an-emerging-threat-as-the-ocean-warms\n",
    "\n",
    "Extrem bloom in the Gulf of Maine in 2016: (ocean warming) https://www.sciencedirect.com/science/article/pii/S0924796322000380\n",
    "\n",
    "\n",
    "### Metabolome of Pseudo-nitzschia and its microbiome\n",
    "\n",
    "https://ami-journals.onlinelibrary.wiley.com/doi/full/10.1111/1462-2920.16242"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome_Id final names</th>\n",
       "      <th>Genome_Id proteins</th>\n",
       "      <th>Genome_Id (beta-version)</th>\n",
       "      <th>Niche_identification</th>\n",
       "      <th>METDB_match_98percent</th>\n",
       "      <th>METDB_taxonomy</th>\n",
       "      <th>Database</th>\n",
       "      <th>Source</th>\n",
       "      <th>RNAPOL_BLAST_IDENTITY</th>\n",
       "      <th>Novelty_Score_Try</th>\n",
       "      <th>...</th>\n",
       "      <th>area_2090_no_prob</th>\n",
       "      <th>percent_area_2090_no_prob</th>\n",
       "      <th>Diff</th>\n",
       "      <th>Mean_Fe</th>\n",
       "      <th>Mean_NO3</th>\n",
       "      <th>Mean_Phos</th>\n",
       "      <th>Mean_Sal</th>\n",
       "      <th>Mean_Si</th>\n",
       "      <th>Mean_SI_NO3</th>\n",
       "      <th>Mean_T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>TARA_ARC_108_MAG_00212</td>\n",
       "      <td>TARA_ARC_MAG_00212</td>\n",
       "      <td>TARA_ARC_108_MAG_00212</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TARA_SMAGs</td>\n",
       "      <td>MAG</td>\n",
       "      <td>83.815333</td>\n",
       "      <td>16.184667</td>\n",
       "      <td>...</td>\n",
       "      <td>8.045617e+05</td>\n",
       "      <td>0.002356</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000948</td>\n",
       "      <td>10.081563</td>\n",
       "      <td>0.987109</td>\n",
       "      <td>33.608312</td>\n",
       "      <td>18.948550</td>\n",
       "      <td>0.354664</td>\n",
       "      <td>3.563677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>TARA_ARC_108_MAG_00228</td>\n",
       "      <td>TARA_ARC_MAG_00228</td>\n",
       "      <td>TARA_ARC_108_MAG_00228</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TARA_SMAGs</td>\n",
       "      <td>MAG</td>\n",
       "      <td>88.626500</td>\n",
       "      <td>11.373500</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>TARA_ARC_108_MAG_00230</td>\n",
       "      <td>TARA_ARC_MAG_00230</td>\n",
       "      <td>TARA_ARC_108_MAG_00230</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TARA_SMAGs</td>\n",
       "      <td>MAG</td>\n",
       "      <td>86.164667</td>\n",
       "      <td>13.835333</td>\n",
       "      <td>...</td>\n",
       "      <td>5.540776e+06</td>\n",
       "      <td>0.016227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001666</td>\n",
       "      <td>2.997443</td>\n",
       "      <td>1.141365</td>\n",
       "      <td>29.205857</td>\n",
       "      <td>11.861578</td>\n",
       "      <td>0.368599</td>\n",
       "      <td>-0.430591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>TARA_IOS_50_MAG_00115</td>\n",
       "      <td>TARA_IOS_MAG_00115</td>\n",
       "      <td>TARA_IOS_50_MAG_00115</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TARA_SMAGs</td>\n",
       "      <td>MAG</td>\n",
       "      <td>88.085000</td>\n",
       "      <td>11.915000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>TARA_PSE_93_MAG_00213</td>\n",
       "      <td>TARA_PSE_MAG_00213</td>\n",
       "      <td>TARA_PSE_93_MAG_00213</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TARA_SMAGs</td>\n",
       "      <td>MAG</td>\n",
       "      <td>87.092000</td>\n",
       "      <td>12.908000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.253203e+07</td>\n",
       "      <td>0.065989</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000693</td>\n",
       "      <td>7.193099</td>\n",
       "      <td>0.847105</td>\n",
       "      <td>35.199538</td>\n",
       "      <td>6.079463</td>\n",
       "      <td>0.289565</td>\n",
       "      <td>22.591633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>TARA_PSW_86_MAG_00222</td>\n",
       "      <td>TARA_PSW_MAG_00222</td>\n",
       "      <td>TARA_PSW_86_MAG_00222</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TARA_SMAGs</td>\n",
       "      <td>MAG</td>\n",
       "      <td>85.232500</td>\n",
       "      <td>14.767500</td>\n",
       "      <td>...</td>\n",
       "      <td>8.352864e+06</td>\n",
       "      <td>0.024463</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.010638</td>\n",
       "      <td>0.191910</td>\n",
       "      <td>36.139248</td>\n",
       "      <td>0.769953</td>\n",
       "      <td>0.018629</td>\n",
       "      <td>24.277400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604</th>\n",
       "      <td>TARA_PSW_86_MAG_00236</td>\n",
       "      <td>TARA_PSW_MAG_00236</td>\n",
       "      <td>TARA_PSW_86_MAG_00236</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TARA_SMAGs</td>\n",
       "      <td>MAG</td>\n",
       "      <td>86.361333</td>\n",
       "      <td>13.638667</td>\n",
       "      <td>...</td>\n",
       "      <td>6.608633e+06</td>\n",
       "      <td>0.019355</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000710</td>\n",
       "      <td>6.436792</td>\n",
       "      <td>0.784656</td>\n",
       "      <td>35.235364</td>\n",
       "      <td>5.522014</td>\n",
       "      <td>0.271339</td>\n",
       "      <td>23.048863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1075</th>\n",
       "      <td>METDB_00365</td>\n",
       "      <td>METDB_00365</td>\n",
       "      <td>METDB_00365</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>METdb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1076</th>\n",
       "      <td>METDB_00366</td>\n",
       "      <td>METDB_00366</td>\n",
       "      <td>METDB_00366</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>METdb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1077</th>\n",
       "      <td>METDB_00367</td>\n",
       "      <td>METDB_00367</td>\n",
       "      <td>METDB_00367</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>METdb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1078</th>\n",
       "      <td>METDB_00368</td>\n",
       "      <td>METDB_00368</td>\n",
       "      <td>METDB_00368</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>METdb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1079</th>\n",
       "      <td>METDB_00369</td>\n",
       "      <td>METDB_00369</td>\n",
       "      <td>METDB_00369</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>METdb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1080</th>\n",
       "      <td>METDB_00370</td>\n",
       "      <td>METDB_00370</td>\n",
       "      <td>METDB_00370</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>METdb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081</th>\n",
       "      <td>METDB_00371</td>\n",
       "      <td>METDB_00371</td>\n",
       "      <td>METDB_00371</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>METdb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1082</th>\n",
       "      <td>METDB_00372</td>\n",
       "      <td>METDB_00372</td>\n",
       "      <td>METDB_00372</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>METdb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 91 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Genome_Id final names  Genome_Id proteins Genome_Id (beta-version)  \\\n",
       "168   TARA_ARC_108_MAG_00212  TARA_ARC_MAG_00212   TARA_ARC_108_MAG_00212   \n",
       "180   TARA_ARC_108_MAG_00228  TARA_ARC_MAG_00228   TARA_ARC_108_MAG_00228   \n",
       "182   TARA_ARC_108_MAG_00230  TARA_ARC_MAG_00230   TARA_ARC_108_MAG_00230   \n",
       "338    TARA_IOS_50_MAG_00115  TARA_IOS_MAG_00115    TARA_IOS_50_MAG_00115   \n",
       "545    TARA_PSE_93_MAG_00213  TARA_PSE_MAG_00213    TARA_PSE_93_MAG_00213   \n",
       "603    TARA_PSW_86_MAG_00222  TARA_PSW_MAG_00222    TARA_PSW_86_MAG_00222   \n",
       "604    TARA_PSW_86_MAG_00236  TARA_PSW_MAG_00236    TARA_PSW_86_MAG_00236   \n",
       "1075             METDB_00365         METDB_00365              METDB_00365   \n",
       "1076             METDB_00366         METDB_00366              METDB_00366   \n",
       "1077             METDB_00367         METDB_00367              METDB_00367   \n",
       "1078             METDB_00368         METDB_00368              METDB_00368   \n",
       "1079             METDB_00369         METDB_00369              METDB_00369   \n",
       "1080             METDB_00370         METDB_00370              METDB_00370   \n",
       "1081             METDB_00371         METDB_00371              METDB_00371   \n",
       "1082             METDB_00372         METDB_00372              METDB_00372   \n",
       "\n",
       "     Niche_identification METDB_match_98percent METDB_taxonomy    Database  \\\n",
       "168                   Yes                   NaN            NaN  TARA_SMAGs   \n",
       "180                    No                   NaN            NaN  TARA_SMAGs   \n",
       "182                   Yes                   NaN            NaN  TARA_SMAGs   \n",
       "338                    No                   NaN            NaN  TARA_SMAGs   \n",
       "545                   Yes                   NaN            NaN  TARA_SMAGs   \n",
       "603                   Yes                   NaN            NaN  TARA_SMAGs   \n",
       "604                   Yes                   NaN            NaN  TARA_SMAGs   \n",
       "1075                  NaN                   NaN            NaN       METdb   \n",
       "1076                  NaN                   NaN            NaN       METdb   \n",
       "1077                  NaN                   NaN            NaN       METdb   \n",
       "1078                  NaN                   NaN            NaN       METdb   \n",
       "1079                  NaN                   NaN            NaN       METdb   \n",
       "1080                  NaN                   NaN            NaN       METdb   \n",
       "1081                  NaN                   NaN            NaN       METdb   \n",
       "1082                  NaN                   NaN            NaN       METdb   \n",
       "\n",
       "     Source  RNAPOL_BLAST_IDENTITY  Novelty_Score_Try  ... area_2090_no_prob  \\\n",
       "168     MAG              83.815333          16.184667  ...      8.045617e+05   \n",
       "180     MAG              88.626500          11.373500  ...               NaN   \n",
       "182     MAG              86.164667          13.835333  ...      5.540776e+06   \n",
       "338     MAG              88.085000          11.915000  ...               NaN   \n",
       "545     MAG              87.092000          12.908000  ...      2.253203e+07   \n",
       "603     MAG              85.232500          14.767500  ...      8.352864e+06   \n",
       "604     MAG              86.361333          13.638667  ...      6.608633e+06   \n",
       "1075    NaN                    NaN                NaN  ...               NaN   \n",
       "1076    NaN                    NaN                NaN  ...               NaN   \n",
       "1077    NaN                    NaN                NaN  ...               NaN   \n",
       "1078    NaN                    NaN                NaN  ...               NaN   \n",
       "1079    NaN                    NaN                NaN  ...               NaN   \n",
       "1080    NaN                    NaN                NaN  ...               NaN   \n",
       "1081    NaN                    NaN                NaN  ...               NaN   \n",
       "1082    NaN                    NaN                NaN  ...               NaN   \n",
       "\n",
       "     percent_area_2090_no_prob  Diff   Mean_Fe   Mean_NO3 Mean_Phos  \\\n",
       "168                   0.002356   NaN  0.000948  10.081563  0.987109   \n",
       "180                        NaN   NaN       NaN        NaN       NaN   \n",
       "182                   0.016227   NaN  0.001666   2.997443  1.141365   \n",
       "338                        NaN   NaN       NaN        NaN       NaN   \n",
       "545                   0.065989   NaN  0.000693   7.193099  0.847105   \n",
       "603                   0.024463   NaN  0.000009   0.010638  0.191910   \n",
       "604                   0.019355   NaN  0.000710   6.436792  0.784656   \n",
       "1075                       NaN   NaN       NaN        NaN       NaN   \n",
       "1076                       NaN   NaN       NaN        NaN       NaN   \n",
       "1077                       NaN   NaN       NaN        NaN       NaN   \n",
       "1078                       NaN   NaN       NaN        NaN       NaN   \n",
       "1079                       NaN   NaN       NaN        NaN       NaN   \n",
       "1080                       NaN   NaN       NaN        NaN       NaN   \n",
       "1081                       NaN   NaN       NaN        NaN       NaN   \n",
       "1082                       NaN   NaN       NaN        NaN       NaN   \n",
       "\n",
       "       Mean_Sal    Mean_Si Mean_SI_NO3     Mean_T  \n",
       "168   33.608312  18.948550    0.354664   3.563677  \n",
       "180         NaN        NaN         NaN        NaN  \n",
       "182   29.205857  11.861578    0.368599  -0.430591  \n",
       "338         NaN        NaN         NaN        NaN  \n",
       "545   35.199538   6.079463    0.289565  22.591633  \n",
       "603   36.139248   0.769953    0.018629  24.277400  \n",
       "604   35.235364   5.522014    0.271339  23.048863  \n",
       "1075        NaN        NaN         NaN        NaN  \n",
       "1076        NaN        NaN         NaN        NaN  \n",
       "1077        NaN        NaN         NaN        NaN  \n",
       "1078        NaN        NaN         NaN        NaN  \n",
       "1079        NaN        NaN         NaN        NaN  \n",
       "1080        NaN        NaN         NaN        NaN  \n",
       "1081        NaN        NaN         NaN        NaN  \n",
       "1082        NaN        NaN         NaN        NaN  \n",
       "\n",
       "[15 rows x 91 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "tara_euk = pd.read_excel(\"genomes/TARA_euk_genomes/SMAGS_supplemental/1-s2.0-S2666979X22000477-mmc5.xlsx\")\n",
    "tara_euk[tara_euk[\"Best_taxonomy_GENRE\"] == \"Pseudo-nitzschia\"]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This genome: TARA_ARC_108_MAG_00230 is found only in the artic so it should be a good candidate to search for bacterial genomes since all samples from the same region. On the other hand, maybe is a weird strain of Pseudo-nitzschia that is only found in the artic and not in other regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alteromonas\n",
      "Series([], Name: user_genome, dtype: object)\n",
      "Series([], Name: Species gtdbtk, dtype: object)\n",
      "Marinobacter\n",
      "1373    TARA_PSE_93_MAG_00042\n",
      "Name: user_genome, dtype: object\n",
      "1373    Marinobacterium litorale\n",
      "Name: Species gtdbtk, dtype: object\n",
      "Polaribacter\n",
      "1430    TARA_PSE_93_MAG_00136\n",
      "Name: user_genome, dtype: object\n",
      "1430    NaN\n",
      "Name: Species gtdbtk, dtype: object\n",
      "Sulfitobacter\n",
      "1417    TARA_PSE_93_MAG_00115\n",
      "Name: user_genome, dtype: object\n",
      "1417    Sulfitobacter pontiacus\n",
      "Name: Species gtdbtk, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# Marinobacter, Alteromonas, Polaribacter.\n",
    "station_id = \"TARA_PSE_93\" # Marinobacter, Polaribacter\n",
    "\n",
    "taxa = [\"Alteromonas\", \"Marinobacter\", \"Polaribacter\", \"Sulfitobacter\"]\n",
    "\n",
    "tara_pro = pd.read_excel(\"genomes/TARA_pro_genomes/MAGS_supplemental/Table_S02_statistics_1888_MAGs.xlsx\", sheet_name=\"gtdbk\")\n",
    "pse_93_pro = tara_pro[tara_pro.user_genome.str.contains(station_id)].dropna(subset=[\"Genus gtdbtk\"])\n",
    "\n",
    "for taxon in taxa:\n",
    "    hits = pse_93_pro.user_genome[pse_93_pro[\"Genus gtdbtk\"].str.contains(taxon)]\n",
    "    taxon_hits = pse_93_pro[\"Species gtdbtk\"][pse_93_pro[\"Genus gtdbtk\"].str.contains(taxon)]\n",
    "    print(taxon)\n",
    "    print(hits)\n",
    "    print(taxon_hits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Matches: Polaribacter (Flavobacteriaceae, also Dokdonia), Marinobacterium (Alteromonadaceae, same as Marinobacter and Alteromonas)\n",
    "\n",
    "# Alternomonas and Marinobacterium also among the MAGs generate from the diatom exudate experiment. Polaribacter found in the dinoflagellate exudate experiment.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alteromonas\n",
      "Series([], Name: user_genome, dtype: object)\n",
      "Series([], Name: Species gtdbtk, dtype: object)\n",
      "Marinobacter\n",
      "1667    TARA_PSW_86_MAG_00232\n",
      "Name: user_genome, dtype: object\n",
      "1667    Marinobacter adhaerens\n",
      "Name: Species gtdbtk, dtype: object\n",
      "Polaribacter\n",
      "Series([], Name: user_genome, dtype: object)\n",
      "Series([], Name: Species gtdbtk, dtype: object)\n",
      "Sulfitobacter\n",
      "Series([], Name: user_genome, dtype: object)\n",
      "Series([], Name: Species gtdbtk, dtype: object)\n"
     ]
    }
   ],
   "source": [
    "station_id = \"TARA_PSW_86\" # Marinobacter\n",
    "\n",
    "taxa = [\"Alteromonas\", \"Marinobacter\", \"Polaribacter\", \"Sulfitobacter\"]\n",
    "\n",
    "tara_pro = pd.read_excel(\"genomes/TARA_pro_genomes/MAGS_supplemental/Table_S02_statistics_1888_MAGs.xlsx\", sheet_name=\"gtdbk\")\n",
    "pse_93_pro = tara_pro[tara_pro.user_genome.str.contains(station_id)].dropna(subset=[\"Genus gtdbtk\"])\n",
    "\n",
    "for taxon in taxa:\n",
    "    hits = pse_93_pro.user_genome[pse_93_pro[\"Genus gtdbtk\"].str.contains(taxon)]\n",
    "    taxon_hits = pse_93_pro[\"Species gtdbtk\"][pse_93_pro[\"Genus gtdbtk\"].str.contains(taxon)]\n",
    "    print(taxon)\n",
    "    print(hits)\n",
    "    print(taxon_hits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alteromonas\n",
      "736    TARA_IOS_50_MAG_00117\n",
      "Name: user_genome, dtype: object\n",
      "736    Alteromonas macleodii\n",
      "Name: Species gtdbtk, dtype: object\n",
      "Marinobacter\n",
      "735    TARA_IOS_50_MAG_00116\n",
      "Name: user_genome, dtype: object\n",
      "735    NaN\n",
      "Name: Species gtdbtk, dtype: object\n",
      "Polaribacter\n",
      "Series([], Name: user_genome, dtype: object)\n",
      "Series([], Name: Species gtdbtk, dtype: object)\n",
      "Sulfitobacter\n",
      "Series([], Name: user_genome, dtype: object)\n",
      "Series([], Name: Species gtdbtk, dtype: object)\n"
     ]
    }
   ],
   "source": [
    "station_id = \"TARA_IOS_50\"  # Alteromonas, Marinobacter\n",
    "\n",
    "taxa = [\"Alteromonas\", \"Marinobacter\", \"Polaribacter\", \"Sulfitobacter\"]\n",
    "\n",
    "tara_pro = pd.read_excel(\"genomes/TARA_pro_genomes/MAGS_supplemental/Table_S02_statistics_1888_MAGs.xlsx\", sheet_name=\"gtdbk\")\n",
    "pse_93_pro = tara_pro[tara_pro.user_genome.str.contains(station_id)].dropna(subset=[\"Genus gtdbtk\"])\n",
    "\n",
    "for taxon in taxa:\n",
    "    hits = pse_93_pro.user_genome[pse_93_pro[\"Genus gtdbtk\"].str.contains(taxon)]\n",
    "    taxon_hits = pse_93_pro[\"Species gtdbtk\"][pse_93_pro[\"Genus gtdbtk\"].str.contains(taxon)]\n",
    "    print(taxon)\n",
    "    print(hits)\n",
    "    print(taxon_hits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alteromonas\n",
      "420    TARA_ARC_108_MAG_00080\n",
      "509    TARA_ARC_108_MAG_00179\n",
      "Name: user_genome, dtype: object\n",
      "420    Alteromonas\n",
      "509    Alteromonas\n",
      "Name: Genus gtdbtk, dtype: object\n",
      "Marinobacter\n",
      "504    TARA_ARC_108_MAG_00174\n",
      "Name: user_genome, dtype: object\n",
      "504    Marinobacter\n",
      "Name: Genus gtdbtk, dtype: object\n",
      "Polaribacter\n",
      "527    TARA_ARC_108_MAG_00201\n",
      "Name: user_genome, dtype: object\n",
      "527    Polaribacter\n",
      "Name: Genus gtdbtk, dtype: object\n",
      "Sulfitobacter\n",
      "423    TARA_ARC_108_MAG_00083\n",
      "454    TARA_ARC_108_MAG_00117\n",
      "472    TARA_ARC_108_MAG_00139\n",
      "Name: user_genome, dtype: object\n",
      "423      Sulfitobacter\n",
      "454    Sulfitobacter_C\n",
      "472      Sulfitobacter\n",
      "Name: Genus gtdbtk, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# NOTE: this is not the TARA station ID but the ocean region. In this case, the arctic, with 108 total metagenomes coassembled to generated these MAGs.\n",
    "\n",
    "station_id = \"TARA_ARC_108\"  # Alteromonas (2), Marinobacter, Polaribacter\n",
    "taxa = [\"Alteromonas\", \"Marinobacter\", \"Polaribacter\", \"Sulfitobacter\"]\n",
    "\n",
    "tara_pro = pd.read_excel(\"genomes/TARA_pro_genomes/MAGS_supplemental/Table_S02_statistics_1888_MAGs.xlsx\", sheet_name=\"gtdbk\")\n",
    "pse_93_pro = tara_pro[tara_pro.user_genome.str.contains(station_id)].dropna(subset=[\"Genus gtdbtk\"])\n",
    "\n",
    "for taxon in taxa:\n",
    "    hits = pse_93_pro.user_genome[pse_93_pro[\"Genus gtdbtk\"].str.contains(taxon)]\n",
    "    taxon_hits = pse_93_pro[\"Genus gtdbtk\"][pse_93_pro[\"Genus gtdbtk\"].str.contains(taxon)]\n",
    "    print(taxon)\n",
    "    print(hits)\n",
    "    print(taxon_hits)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pseudo-nitzschia MAGs in the TARA dataset:\n",
    "\n",
    "Pseudo-nitzschia: TARA_ARC_108_MAG_00212 (Proteins: TARA_ARC_MAG_00212) or TARA_ARC_108_MAG_00230 (Proteins: TARA_ARC_MAG_00230)\n",
    "\n",
    "Alteromonas\n",
    "TARA_ARC_108_MAG_00080\n",
    "TARA_ARC_108_MAG_00179\n",
    "\n",
    "Marinobacter\n",
    "TARA_ARC_108_MAG_00174\n",
    "\n",
    "Polaribacter\n",
    "TARA_ARC_108_MAG_00201\n",
    "\n",
    "Sulfitobacter sp.\n",
    "TARA_ARC_108_MAG_00083\n",
    "TARA_ARC_108_MAG_00117\n",
    "TARA_ARC_108_MAG_00139\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tarfile\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "euk_genomes = [\n",
    "    'TARA_ARC_108_MAG_00212', 'TARA_ARC_108_MAG_00230'\n",
    "    ]\n",
    "prok_genomes = [\n",
    "    'TARA_ARC_108_MAG_00080', 'TARA_ARC_108_MAG_00179',\n",
    "    'TARA_ARC_108_MAG_00174', 'TARA_ARC_108_MAG_00201'\n",
    "]\n",
    "\n",
    "target_dir = Path('genomes/pseudo-nitzschia')\n",
    "target_dir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "# Eukaryotic genomes\n",
    "tar_file = 'genomes/TARA_euk_genomes/SMAGs_contigs_individual.fna.tar.gz'\n",
    "files_to_extract = [f'{genome}.fa' for genome in euk_genomes]\n",
    "\n",
    "with tarfile.open(tar_file, 'r') as tar:\n",
    "    for member in tar.getmembers():\n",
    "        if any(name in member.name for name in files_to_extract):\n",
    "            member.name = Path(member.name).name\n",
    "            tar.extract(member, path=target_dir)\n",
    "\n",
    "# Prokaryotic genomes\n",
    "tar_file = 'genomes/TARA_pro_genomes/FASTA_1888_MAGs_Bac_Arc.tar.gz'\n",
    "files_to_extract = [f'{genome}.fa' for genome in prok_genomes]\n",
    "\n",
    "with tarfile.open(tar_file, 'r') as tar:\n",
    "    for member in tar.getmembers():\n",
    "        if any(name in member.name for name in files_to_extract):\n",
    "            member.name = Path(member.name).name\n",
    "            tar.extract(member, path=target_dir)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sources of algal genomes\n",
    "\n",
    "[Phycocosm](https://phycocosm.jgi.doe.gov/phycocosm/home)\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Candidate objective functions\n",
    "\n",
    "Candidate objective functions to explore, which may actually require full community modelling, rather than only the phycosphere:\n",
    "\n",
    "1. Metabolic Dominance Detection: This function aims to quantify the metabolic dominance of the phycosphere during the early stages of a HAB event. It could be defined as the ratio of the total metabolic flux of the dominant HAB species to the total metabolic flux of the entire microbial community. A significant increase in this ratio could indicate the onset of a HAB event.\n",
    "\n",
    "2. Early Warning Signal Detection: This function aims to identify drastic shifts in metabolic fluxes that could serve as early warning signals for a HAB event. It could be defined as the rate of change of the total metabolic flux of the microbial community, with a significant increase indicating a potential HAB event.\n",
    "\n",
    "3. Ecological Health Metric Prototyping: This function aims to develop ecological health metrics that can effectively monitor coastal ecosystems. It could be defined as a composite score of the above two functions, with a significant deviation from the baseline indicating a potential threat to the ecological health of the ecosystem.\n",
    "\n",
    "But we need to iterate upon this several times, and also brainstorm ideas with the rest of the sci-team!"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It looks like a large part of the interactions between bacteria and microalgae at the onset of a HAB event are related to cofactors/vitamins and iron or other trace elements in exchange of DOC [1, 2]. Thus, I would expect a shift in metabolic fluxes favoring the production of secreted DOCs by the microalgae [3, 4] and a marked increase in the production and secretion of cofactors/vitamins and siderophores by the bacterial community, as well as the pathways involved in uptaking and processing the diverse DOCs. Other studies have pointed to Quorum Sensing-related pathways during HAB events [5]. These changes could serve as a metabolic, system-wise, signal to detect the onset of a HAB event.\n",
    "\n",
    "We could integrate metatranscriptomic datasets of HAB events [6, 7] into the community/phycosphere metabolic model to find the closest metabolic network state matching the metatranscriptome (this probably will be a collection of optimal states, think of flux sampling constrained by transcript data). By comparing these states with baseline states (e.g, metaTs outside of a HAB event) we could learn community-wide metabolic signatures that are characteristic of the onset of a HAB event.\n",
    "\n",
    "\n",
    "Refs:\n",
    "\n",
    "1. https://www.pnas.org/doi/full/10.1073/pnas.1009566107\n",
    "2. https://www.pnas.org/doi/full/10.1073/pnas.0905512106 (iron)\n",
    "3. https://www.sciencedirect.com/science/article/pii/S0043135420300907\n",
    "4. https://journals.asm.org/doi/10.1128/MRA.00724-20\n",
    "5. https://www.sciencedirect.com/science/article/pii/S0944501318306189\n",
    "6. https://www.sciencedirect.com/science/article/pii/S1568988314000729\n",
    "7. https://www.nature.com/articles/s41598-020-63326-8"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pseudo-Nitzschia "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2023-11-05\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>TARA_ARC_108_MAG_00212_cds</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>7fd8300b2df0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>2227</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>4762</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>1957</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>359</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*bof_c - 1.0*bof_c_reverse_660d5</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>c, u, e, h</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model TARA_ARC_108_MAG_00212_cds at 0x7fd8300b2df0>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import cobra\n",
    "\n",
    "model = cobra.io.read_sbml_model(\"gems/photoeukstein/TARA_ARC_108_MAG_00212.cds.xml.gz\")\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 33.330</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ATPCS_c</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CISY_m</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PDH_E1a_h</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADH_c</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAPDH_h</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PIt_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PINAt_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_h_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_h2o_e</th>\n",
       "      <td>370.974157</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PREPHENATEDEHYDRAT-RXN</th>\n",
       "      <td>-0.001000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4762 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 33.330 at 0x7fd7fdeac280>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.optimize()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Domoic Acid Pathway\n",
    "\n",
    "it is barely characterized: https://www.science.org/doi/10.1126/science.aau0382\n",
    "\n",
    "https://www.eurekaselect.com/article/124778"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Metabolite identifier</strong></td><td>glu__L_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>L-Glutamate</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x7fd7d80cc5e0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Formula</strong></td><td>C5H8NO4</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartment</strong></td><td>c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 44 reaction(s)</strong></td><td>\n",
       "                    PRFGS, GACT, GLUS, GLUTL, GLUNpp, MTHF3ES_c, AAAT, AOTA, biomass_rxn_okamuranus, POAT, RXN-12878, CHRL_h, AOOAh, OAT_m, CBPS, GDHm, GLUNA1t_h, GLUTAMATE-DEHYDROGENASE-NADP+-RXN, P5CDH_m, GLURS-RXN,...\n",
       "                    </td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Metabolite glu__L_c at 0x7fd7d80cc5e0>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites.get_by_id(\"glu__L_c\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Metabolite identifier</strong></td><td>grdp_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Geranyl diphosphate</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x7fd806de7040</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Formula</strong></td><td>C10H17O7P2</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartment</strong></td><td>c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 2 reaction(s)</strong></td><td>\n",
       "                    FPPS, GPPS\n",
       "                    </td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Metabolite grdp_c at 0x7fd806de7040>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[met for met in model.metabolites if \"geranyl\" in met.name.lower()][0]  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>FPPS</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>farnesyl pyrophosphate synthase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x7fd805f40250</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>grdp_c + ipdp_c --> frdp_c + ppi_c</p>\n",
       "                    <p style='text-align:right'>Geranyl diphosphate + Isopentenyl diphosphate --> trans,trans-Farnesyl diphosphate + Diphosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>TARA_ARC_108_MAG_00212_000000000657_6_1 or TARA_ARC_108_MAG_00212_000000002191_11_1 or...</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction FPPS at 0x7fd805f40250>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id(\"FPPS\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GPPS</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>geranyl pyrophosphate synthase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x7fd805edef70</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>dmpp_c + ipdp_c --> grdp_c + ppi_c</p>\n",
       "                    <p style='text-align:right'>Dimethylallyl diphosphate + Isopentenyl diphosphate --> Geranyl diphosphate + Diphosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>TARA_ARC_108_MAG_00212_000000000155_17_1 or TARA_ARC_108_MAG_00212_000000000657_6_1 or...</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GPPS at 0x7fd805edef70>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id(\"GPPS\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alteromonas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>MAG_00080_bigg</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>7fd7cf4c5370</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>2184</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>3115</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>816</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*BIOMASS_reaction - 1.0*BIOMASS_reaction_reverse_5a818</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytoplasm, extracellular, periplasm, mitochondrion, unknown, golgi, nucleus, lysosome, chloroplast, mitochondrial intermembrane space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model MAG_00080_bigg at 0x7fd7cf4c5370>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = cobra.io.read_sbml_model(\"gems/MAG_00080_bigg.xml\")\n",
    "model"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "humann",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
